src/js/plugin.js

```diff
@@ -0,0 +1,342 @@
+ *
+ * This factory function will replace itself with an instance of the requested
+ * sub-class of Plugin.
+ *
+ * @private
+ * @return {Function}
+ */
+const createPluginFactory = (name, PluginSubClass) => {
+
+  // Add a `name` property to the plugin prototype so that each plugin can
+  // refer to itself by name.
+  PluginSubClass.prototype.name = name;
+
+  return function(...args) {
+    this[name] = new PluginSubClass(...[this, ...args]);
+    return this[name];
+  };
+};
+
+class Plugin {
+
+  /**
+   * Plugin constructor.
+   *
+   * Subclasses should make sure they call `super` in order to make sure their
+   * plugins are properly initialized.
+   *
+   * @param {Player} player
+   */
+  constructor(player) {
+    this.player = player;
+    evented(this, {exclude: ['trigger']});
+    stateful(this, this.constructor.defaultState);
+    markPluginAsActive(player, this.name);
+    player.one('dispose', Fn.bind(this, this.dispose));
+    player.trigger('pluginsetup', this.getEventHash_());
+  }
+
+  /**
+   * Each event triggered by plugins includes a hash of additional data with
+   * conventional properties.
+   *
+   * This returns that object or mutates an existing hash.
+   *
+   * @private
+   * @param  {Object} [hash={}]
+   * @return {Object}
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   */
+  getEventHash_(hash = {}) {
+    hash.name = this.name;
+    hash.plugin = this.constructor;
+    hash.instance = this;
+    return hash;
+  }
+
+  /**
+   * Triggers an event on the plugin object.
+   *
+   * @param  {Event|Object|string} event
+   *         A string (the type) or an event object with a type attribute.
+   *
+   * @param  {Object} [hash={}]
+   *         Additional data hash to pass along with the event. In this case,
+   *         several properties are added to the hash:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   *
+   * @return {boolean}
+   *         Whether or not default was prevented.
+   */
+  trigger(event, hash = {}) {
+    return Events.trigger(this.eventBusEl_, event, this.getEventHash_(hash));
+  }
+
+  /**
+   * Disposes a plugin.
+   *
+   * Subclasses can override this if they want, but for the sake of safety,
+   * it's probably best to subscribe to one of the disposal events.
+   */
+  dispose() {
+    const {name, player} = this;
+
+    this.trigger('dispose');
+    this.off();
+
+    // Eliminate any possible sources of leaking memory by clearing up references
+    // between the player and the plugin instance and nulling out the plugin's
+    // state and replacing methods with a function that throws.
+    player[PLUGIN_CACHE_KEY][name] = false;
+    this.player = this.state = null;
+
+    this.dispose = () => {
```

Comment:
This seems like it is too much protection for the user, it could be overly restrictive to some plugins.